vkubelet/podcontroller.go

```diff
@@ -0,0 +1,380 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package vkubelet
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/cpuguy83/strongerrors/status/ocstatus"
+	pkgerrors "github.com/pkg/errors"
+	"go.opencensus.io/trace"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/informers/core/v1"
+	"k8s.io/client-go/kubernetes/scheme"
+	typedcorev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	corev1listers "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/workqueue"
+
+	"github.com/virtual-kubelet/virtual-kubelet/log"
+)
+
+const (
+	// maxRetries is the number of times we try to process a given key before permanently forgetting it.
+	maxRetries = 20
+)
+
+// PodController is the controller implementation for Pod resources.
+type PodController struct {
+	// server is the instance to which this controller belongs.
+	server *Server
+	// podsInformer is an informer for Pod resources.
+	podsInformer v1.PodInformer
+	// podsLister is able to list/get Pod resources from a shared informer's store.
+	podsLister corev1listers.PodLister
+
+	// workqueue is a rate limited work queue.
+	// This is used to queue work to be processed instead of performing it as soon as a change happens.
+	// This means we can ensure we only process a fixed amount of resources at a time, and makes it easy to ensure we are never processing the same item simultaneously in two different workers.
+	workqueue workqueue.RateLimitingInterface
+	// recorder is an event recorder for recording Event resources to the Kubernetes API.
+	recorder record.EventRecorder
+
+	// context is the context to use when executing the syncHandler and calling the provider.
+	context context.Context
+}
+
+// NewPodController returns a new instance of PodController.
+func NewPodController(server *Server, context context.Context) *PodController {
+	// Create an event broadcaster.
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(log.L.Infof)
+	eventBroadcaster.StartRecordingToSink(&typedcorev1.EventSinkImpl{Interface: server.k8sClient.CoreV1().Events("")})
+	recorder := eventBroadcaster.NewRecorder(scheme.Scheme, corev1.EventSource{Component: fmt.Sprintf("%s/pod-controller", server.nodeName)})
+
+	// Create an instance of PodController having a work queue that uses the rate limiter created above.
+	pc := &PodController{
+		server:       server,
+		podsInformer: server.podInformer,
+		podsLister:   server.podInformer.Lister(),
+		workqueue:    workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "pods"),
+		recorder:     recorder,
+		context:      context,
```

Comment:
Ideally we wouldn't store the context here but rather pass it into `Run`

Reply:
Indeed! My bad. Context should never be stored and only passed.